content/posts/2019-08-03---Going-from-ember-to-vue.md

```diff
@@ -162,10 +175,16 @@ This means that child component does not need to receive an action as a paramete
 All it does is emit something, pass any properties along with it and then it stops caring.
 It's up to the parent component to listen to it and respond.
 
-One of the things I disliked with the actions helper in Ember was when the child component could call `this.passedAction()` and I would get an error that it wasn't a function because I forgot to pass it down, the child component stopped working.
-With emitting, the child component can continue to function.
-To be clear: making API calls to your backend should happen through the Vuex store -- mutating applicaton state should not be done with `emit`.
-However, if you have buttons component that toggles something on and off, this is a good case for using `emit` as the parent component is most likely the only thing that cares about the event.
+One of the things I disliked with the actions helper in Ember was that a child component might not always want to bubble something up.
+Say you had a component like this in Ember: `<AuthorInfo @updateProfile={{action "updateProfile"}} />`.
+We pass down an action called `updateProfile` which allows the logged in user to update their profile.
+However, I want to use that same layout elsewhere when viewing a post, but it isn't editable.
+So, I simply don't pass down that action: `<AuthorInfo />`.
+If the user triggers that action, however, I will get an error like `this.updateProfile is not a function` (or something similar).
+So, you wrap stuff in if conditions to make sure that's not the case.
+
+With emitting, the child component won't hit this error, as all it does is emit something out and it's up to the parent to intercept it.
```

Comment:
what if you want to be told that you are using an incorrect function? how do you debug this? 
it sounds like it just fails silently, and if you don't have tests over this particular area, you won't know that you maybe have a typo or something

Reply:
You can still pass functions as props to a child component (although it is usually a sign that you might be doing something wrong): https://michaelnthiessen.com/pass-function-as-prop/

For a child component communicating something up, I'd say that failing silently is fine. A component can continue to function even if the right event isn't triggered up the chain.

Reply:
>  I'd say that failing silently is fine.

wouldn't that be hard on maintenance? how would you track things down? like, if you're expecting a parent component or something to react to something that was emitted?

Reply:
If you have a parent that is truly dependent on an update triggered by a child, I think that should be handled in Vuex.  The child can dispatch an event to the store, which triggers a mutation, updates our application state and the parent can observe that.

Emitting is better for non-state specific logic. I.e. I click a button in my ButtonComponent and my parent component updates the background color.  That doesn't need to live at the application-state level. The child just says "Hey, I'm clicking the button!" and it shouldn't break just because the parent did know how to respond to it.  The child should continue to function.

Reply:
gotchya -- so your solution to that, is pretty similar redux (as I think you've said elsewhere). 

 this is bringing up so many frontend data-flow philosophy things. I need to figure out how to organize my thoughts about this and right a blog post on it. :)